File: packages/python/scalix/agent/orchestrator.py
"""Multi-agent orchestration — Team and Pipeline patterns."""

from __future__ import annotations

import asyncio
import logging
from typing import Any

from scalix.agent.agent import Agent
from scalix.types import AgentResult, TeamResult

logger = logging.getLogger("scalix.orchestrator")


class Team:
    """Orchestrate multiple agents working together on a task.

    Agents can work sequentially (pipeline), in parallel, or with
    dynamic routing based on the task.

    Supports three workflow modes:
    - "sequential" — agents run in dict insertion order, each receiving
      the previous agent's output
    - "parallel" — all agents run simultaneously on the same prompt,
      outputs are combined
    - "A → B → C" — explicit ordering with arrow syntax

    Args:
        agents: Named dictionary of agents.
        workflow: Execution strategy — "sequential", "parallel", or
                  a custom workflow string like "researcher → analyst → writer".

    Example:
        from scalix import Agent, Team, Tool

        researcher = Agent(model="auto", tools=[Tool.web_search()])
        analyst = Agent(model="auto", tools=[Tool.code_exec()])
        writer = Agent(model="auto")

        team = Team(
            agents={
                "researcher": researcher,
                "analyst": analyst,
                "writer": writer,
            },
            workflow="researcher → analyst → writer",
        )

        result = await team.run("Create a market analysis report")
    """

    def __init__(
        self,
        agents: dict[str, Agent],
        workflow: str = "sequential",
    ) -> None:
        self.agents = agents
        self.workflow = workflow
        self._order = self._parse_workflow(workflow)

    def _parse_workflow(self, workflow: str) -> list[str]:
        """Parse workflow string into an ordered list of agent names."""
        if workflow == "sequential":
            return list(self.agents.keys())
        elif workflow == "parallel":
            return list(self.agents.keys())
        else:
            # Parse arrow syntax: "researcher → analyst → writer"
            # Support both → and ->
            names = [
                name.strip()
                for name in workflow.replace("→", "->").split("->")
            ]
            # Validate all names exist
            for name in names:
                if name not in self.agents:
                    raise ValueError(
                        f"Agent '{name}' in workflow not found. "
                        f"Available: {list(self.agents.keys())}"
                    )
            return names

    async def run(self, prompt: str, **kwargs: Any) -> TeamResult:
        """Execute the team workflow with the given prompt.

        Args:
            prompt: The initial task prompt.
            **kwargs: Additional parameters passed to each agent.

        Returns:
            TeamResult with combined output from all agents.
        """
        if self.workflow == "parallel":
            return await self._run_parallel(prompt, **kwargs)
        else:
            return await self._run_sequential(prompt, **kwargs)

    async def _run_sequential(self, prompt: str, **kwargs: Any) -> TeamResult:
        """Run agents sequentially, each receiving the previous output."""
        agent_results: dict[str, AgentResult] = {}
        workflow_log: list[str] = []
        current_input = prompt

        for name in self._order:
            agent = self.agents[name]
            logger.debug("Team: running agent '%s'", name)
            workflow_log.append(f"Running agent: {name}")

            result = await agent.run(current_input, **kwargs)
            agent_results[name] = result
            workflow_log.append(f"Agent '{name}' completed: {len(result.output)} chars")

            # Pass this agent's output as input to the next agent
            current_input = (
                f"Previous agent ({name}) produced the following output:\n\n"
                f"{result.output}\n\n"
                f"Original task: {prompt}\n\n"
                f"Continue from where the previous agent left off."
            )

        # Final output is the last agent's output
        final_output = agent_results[self._order[-1]].output if self._order else ""

        return TeamResult(
            output=final_output,
            agent_results=agent_results,
            workflow_log=workflow_log,
        )

    async def _run_parallel(self, prompt: str, **kwargs: Any) -> TeamResult:
        """Run all agents in parallel on the same prompt."""
        agent_results: dict[str, AgentResult] = {}
        workflow_log: list[str] = ["Running all agents in parallel"]

        async def run_agent(name: str, agent: Agent) -> tuple[str, AgentResult]:
            logger.debug("Team (parallel): running agent '%s'", name)
            result = await agent.run(prompt, **kwargs)
            return name, result

        tasks = [
            run_agent(name, self.agents[name])
            for name in self._order
        ]
        results = await asyncio.gather(*tasks, return_exceptions=True)

        for item in results:
            if isinstance(item, BaseException):
                workflow_log.append(f"Agent failed: {item}")
                continue
            assert isinstance(item, tuple)
            name, result = item
            agent_results[name] = result
            workflow_log.append(f"Agent '{name}' completed: {len(result.output)} chars")

        # Combine all outputs
        combined_parts = []
        for name in self._order:
            if name in agent_results:
                combined_parts.append(
                    f"=== {name} ===\n{agent_results[name].output}"
                )

        return TeamResult(
            output="\n\n".join(combined_parts),
            agent_results=agent_results,
            workflow_log=workflow_log,
        )

    def __repr__(self) -> str:
        agent_names = list(self.agents.keys())
        return f"Team(agents={agent_names}, workflow={self.workflow!r})"


class Pipeline:
    """A linear chain of agents where each agent's output feeds the next.

    This is a simpler alternative to Team for straightforward sequential workflows.

    Args:
        steps: Ordered list of agents to execute in sequence.

    Example:
        from scalix import Agent, Pipeline

        pipeline = Pipeline(steps=[
            Agent(instructions="Extract key facts from the input"),
            Agent(instructions="Summarize the facts into bullet points"),
            Agent(instructions="Format as a professional report"),
        ])

        result = await pipeline.run("Raw data: ...")
    """

    def __init__(self, steps: list[Agent]) -> None:
        self.steps = steps

    async def run(self, prompt: str, **kwargs: Any) -> TeamResult:
        """Execute the pipeline sequentially.

        Each agent receives the previous agent's output as its prompt.

        Args:
            prompt: Input to the first agent in the pipeline.
            **kwargs: Additional parameters passed to each agent.

        Returns:
            TeamResult with output from the final agent.
        """
        agent_results: dict[str, AgentResult] = {}
        workflow_log: list[str] = []
        current_input = prompt

        for i, agent in enumerate(self.steps):
            step_name = f"step_{i}"
            logger.debug("Pipeline: running step %d/%d", i + 1, len(self.steps))
            workflow_log.append(f"Running step {i + 1}/{len(self.steps)}")

            result = await agent.run(current_input, **kwargs)
            agent_results[step_name] = result
            workflow_log.append(
                f"Step {i + 1} completed: {len(result.output)} chars"
            )

            # Feed output to next step
            current_input = result.output

        # Final output is the last step's output
        final_output = (
            agent_results[f"step_{len(self.steps) - 1}"].output
            if self.steps
            else ""
        )

        return TeamResult(
            output=final_output,
            agent_results=agent_results,
            workflow_log=workflow_log,
        )

    def __repr__(self) -> str:
        return f"Pipeline(steps={len(self.steps)} agents)"

File: archive/future/typescript/agent/orchestrator.ts
/**
 * Multi-agent orchestration — Team and Pipeline patterns.
 */

import type { Agent } from './agent.js';
import type { AgentResult, TeamResult } from '../types.js';

/**
 * Orchestrate multiple agents working together on a task.
 *
 * Supports three workflow modes:
 * - "sequential" — agents run in insertion order, each receiving
 *   the previous agent's output
 * - "parallel" — all agents run simultaneously on the same prompt,
 *   outputs are combined
 * - "A → B → C" — explicit ordering with arrow syntax
 *
 * @example
 * ```typescript
 * import { Agent, Team, Tool } from 'scalix';
 *
 * const team = new Team({
 *   agents: {
 *     researcher: new Agent({ tools: [Tool.webSearch()] }),
 *     analyst: new Agent({ tools: [Tool.codeExec()] }),
 *     writer: new Agent(),
 *   },
 *   workflow: 'researcher → analyst → writer',
 * });
 *
 * const result = await team.run('Create a market report');
 * ```
 */
export class Team {
  readonly agents: Record<string, Agent>;
  readonly workflow: string;
  private readonly order: string[];

  constructor(options: { agents: Record<string, Agent>; workflow?: string }) {
    this.agents = options.agents;
    this.workflow = options.workflow ?? 'sequential';
    this.order = this.parseWorkflow(this.workflow);
  }

  private parseWorkflow(workflow: string): string[] {
    if (workflow === 'sequential' || workflow === 'parallel') {
      return Object.keys(this.agents);
    }

    // Parse arrow syntax: "researcher → analyst → writer"
    const names = workflow
      .replace(/→/g, '->')
      .split('->')
      .map((s) => s.trim());

    for (const name of names) {
      if (!(name in this.agents)) {
        throw new Error(
          `Agent '${name}' in workflow not found. Available: ${Object.keys(this.agents).join(', ')}`,
        );
      }
    }

    return names;
  }

  async run(prompt: string): Promise<TeamResult> {
    if (this.workflow === 'parallel') {
      return this.runParallel(prompt);
    }
    return this.runSequential(prompt);
  }

  private async runSequential(prompt: string): Promise<TeamResult> {
    const agentResults: Record<string, AgentResult> = {};
    const workflowLog: string[] = [];
    let currentInput = prompt;

    for (const name of this.order) {
      const agent = this.agents[name];
      workflowLog.push(`Running agent: ${name}`);

      const result = await agent.run(currentInput);
      agentResults[name] = result;
      workflowLog.push(`Agent '${name}' completed: ${result.output.length} chars`);

      // Pass this agent's output as input to the next
      currentInput =
        `Previous agent (${name}) produced the following output:\n\n` +
        `${result.output}\n\n` +
        `Original task: ${prompt}\n\n` +
        `Continue from where the previous agent left off.`;
    }

    const finalOutput =
      this.order.length > 0 ? agentResults[this.order[this.order.length - 1]].output : '';

    return { output: finalOutput, agentResults, workflowLog };
  }

  private async runParallel(prompt: string): Promise<TeamResult> {
    const workflowLog: string[] = ['Running all agents in parallel'];

    const entries = await Promise.allSettled(
      this.order.map(async (name) => {
        const result = await this.agents[name].run(prompt);
        return { name, result };
      }),
    );

    const agentResults: Record<string, AgentResult> = {};

    for (const entry of entries) {
      if (entry.status === 'fulfilled') {
        const { name, result } = entry.value;
        agentResults[name] = result;
        workflowLog.push(`Agent '${name}' completed: ${result.output.length} chars`);
      } else {
        workflowLog.push(`Agent failed: ${entry.reason}`);
      }
    }

    // Combine all outputs
    const combinedParts: string[] = [];
    for (const name of this.order) {
      if (name in agentResults) {
        combinedParts.push(`=== ${name} ===\n${agentResults[name].output}`);
      }
    }

    return {
      output: combinedParts.join('\n\n'),
      agentResults,
      workflowLog,
    };
  }
}

/**
 * A linear chain of agents where each agent's output feeds the next.
 *
 * @example
 * ```typescript
 * import { Agent, Pipeline } from 'scalix';
 *
 * const pipeline = new Pipeline([
 *   new Agent({ instructions: 'Extract key facts from the input' }),
 *   new Agent({ instructions: 'Summarize the facts into bullet points' }),
 *   new Agent({ instructions: 'Format as a professional report' }),
 * ]);
 *
 * const result = await pipeline.run('Raw data: ...');
 * ```
 */
export class Pipeline {
  readonly steps: Agent[];

  constructor(steps: Agent[]) {
    this.steps = steps;
  }

  async run(prompt: string): Promise<TeamResult> {
    const agentResults: Record<string, AgentResult> = {};
    const workflowLog: string[] = [];
    let currentInput = prompt;

    for (let i = 0; i < this.steps.length; i++) {
      const stepName = `step_${i}`;
      workflowLog.push(`Running step ${i + 1}/${this.steps.length}`);

      const result = await this.steps[i].run(currentInput);
      agentResults[stepName] = result;
      workflowLog.push(`Step ${i + 1} completed: ${result.output.length} chars`);

      currentInput = result.output;
    }

    const finalOutput =
      this.steps.length > 0 ? agentResults[`step_${this.steps.length - 1}`].output : '';

    return { output: finalOutput, agentResults, workflowLog };
  }
}
